executor/analyze.go

```diff
@@ -496,22 +527,469 @@ type AnalyzeFastTask struct {
+		}
+		e.collectors[hasPKInfo+j].Samples[samplePos].Ordinal = int(samplePos)
+		e.collectors[hasPKInfo+j].Samples[samplePos].Value = v
+	}
+	// Update the indexes' collectors.
+	for j, idxInfo := range e.idxsInfo {
+		idxVals := make([]types.Datum, 0, len(idxInfo.Columns))
+		for _, idxCol := range idxInfo.Columns {
+			for _, colInfo := range e.colsInfo {
+				if colInfo.Name == idxCol.Name {
+					v, err := e.getValueByInfo(colInfo, values)
+					if err != nil {
+						return err
+					}
+					idxVals = append(idxVals, v)
+					break
+				}
+			}
+		}
+		var bytes []byte
+		bytes, err = codec.EncodeKey(e.ctx.GetSessionVars().StmtCtx, bytes, idxVals...)
+		if err != nil {
+			return err
+		}
+		if e.collectors[len(e.colsInfo)+hasPKInfo+j].Samples[samplePos] == nil {
+			e.collectors[len(e.colsInfo)+hasPKInfo+j].Samples[samplePos] = &statistics.SampleItem{}
+		}
+		e.collectors[len(e.colsInfo)+hasPKInfo+j].Samples[samplePos].Ordinal = int(samplePos)
+		e.collectors[len(e.colsInfo)+hasPKInfo+j].Samples[samplePos].Value = types.NewBytesDatum(bytes)
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleBatchGetResponse(kvMap map[string][]byte) (err error) {
+	length := int32(len(kvMap))
+	newCursor := atomic.AddInt32(&e.sampCursor, length)
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	samplePos := newCursor - length
+	for sKey, sValue := range kvMap {
+		err = e.updateCollectorSamples(sValue, kv.Key(sKey), samplePos, hasPKInfo)
+		if err != nil {
+			return err
+		}
+		samplePos++
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleScanIter(iter kv.Iterator) (err error) {
+	hasPKInfo := 0
+	if e.pkInfo != nil {
+		hasPKInfo = 1
+	}
+	rander := rand.New(rand.NewSource(RandSeed + int64(e.rowCount)))
+	for ; iter.Valid() && err == nil; err = iter.Next() {
+		// reservoir sampling
+		e.rowCount++
+		randNum := rander.Int63n(int64(e.rowCount))
+		if randNum > int64(MaxSampleSize) && e.sampCursor == int32(MaxSampleSize) {
+			continue
+		}
+
+		p := rander.Int31n(int32(MaxSampleSize))
+		if e.sampCursor < int32(MaxSampleSize) {
+			p = e.sampCursor
+			e.sampCursor++
+		}
+
+		err = e.updateCollectorSamples(iter.Value(), iter.Key(), p, hasPKInfo)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func (e *AnalyzeFastExec) handleScanTasks(bo *tikv.Backoffer) error {
+	snapshot, err := e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	if err != nil {
+		return err
+	}
+	for _, t := range e.scanTasks {
+		iter, err := snapshot.Iter(t.StartKey, t.EndKey)
+		if err != nil {
+			return err
+		}
+		err = e.handleScanIter(iter)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *AnalyzeFastExec) handleSampTasks(bo *tikv.Backoffer, rid int, err *error) {
+	defer func() {
+		e.wg.Done()
+	}()
+	var snapshot kv.Snapshot
+	snapshot, *err = e.ctx.GetStore().(tikv.Storage).GetSnapshot(kv.MaxVersion)
+	rander := rand.New(rand.NewSource(RandSeed + int64(rid)))
```

Comment:
`RandSeed` and `rid` are both constants whenever we run fast analyze jobs. Which means all the rand number are the same when we run the fast analyze for the same table in different time. Further cause the same statistics.